In [0]:
!nvidia-smi

Fri Jan 24 01:26:01 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.44       Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   37C    P8     9W /  70W |      0MiB / 15079MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU  

In [0]:
import pynvml
pynvml.nvmlInit()
handle = pynvml.nvmlDeviceGetHandleByIndex(0)
device_name = pynvml.nvmlDeviceGetName(handle)
if (device_name != b'Tesla T4') and (device_name != b'Tesla P100-PCIE-16GB'):
  raise Exception("""
    Unfortunately this instance does not have a T4 or P100 GPU.
    
    Please make sure you've configured Colab to request a GPU instance type.
    
    Sometimes Colab allocates a Tesla K80 instead of a T4 or P100. Resetting the instance.
If you get a K80 GPU, try Runtime -> Reset all runtimes...
  """)
else:
  print('Yes, you got the right kind of GPU to work and it is a Tesla T4 GPU.')

Yes, you got the right kind of GPU to work and it is a Tesla T4 GPU.


In [0]:
!wget -nc https://github.com/rapidsai/notebooks-extended/raw/master/utils/rapids-colab.sh
!bash rapids-colab.sh

import sys, os

sys.path.append('/usr/local/lib/python3.6/site-packages/')
os.environ['NUMBAPRO_NVVM'] = '/usr/local/cuda/nvvm/lib64/libnvvm.so'
os.environ['NUMBAPRO_LIBDEVICE'] = '/usr/local/cuda/nvvm/libdevice/'

--2020-01-24 01:26:13--  https://github.com/rapidsai/notebooks-extended/raw/master/utils/rapids-colab.sh
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://github.com/rapidsai/notebooks-contrib/raw/master/utils/rapids-colab.sh [following]
--2020-01-24 01:26:14--  https://github.com/rapidsai/notebooks-contrib/raw/master/utils/rapids-colab.sh
Reusing existing connection to github.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/rapidsai/notebooks-contrib/master/utils/rapids-colab.sh [following]
--2020-01-24 01:26:14--  https://raw.githubusercontent.com/rapidsai/notebooks-contrib/master/utils/rapids-colab.sh
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.co

In [0]:
import seaborn as sns
import matplotlib as mpl
import matplotlib.pyplot as plt
%matplotlib inline

In [0]:
print("matplotlib stype available:\n\n{}".format(plt.style.available))

matplotlib stype available:

['seaborn-paper', 'tableau-colorblind10', 'seaborn-white', 'dark_background', 'seaborn-bright', 'seaborn-ticks', 'seaborn-talk', 'seaborn-colorblind', 'seaborn-darkgrid', 'fivethirtyeight', 'Solarize_Light2', 'seaborn-dark-palette', 'seaborn-deep', 'classic', '_classic_test', 'seaborn-notebook', 'bmh', 'seaborn-muted', 'fast', 'grayscale', 'seaborn-pastel', 'seaborn-poster', 'ggplot', 'seaborn', 'seaborn-dark', 'seaborn-whitegrid']


In [0]:
mpl.style.use(['Solarize_Light2'])
print("Using style of 'Solarize_Light2' to plot")

Using style of 'Solarize_Light2' to plot


In [0]:
import cudf
import cuml

In [0]:
print(f"cuDF's version is: {cudf.__version__}")

cuDF's version is: 0.10.0


In [0]:
print(f"cuML's version is: {cuml.__version__}")

cuML's version is: 0.10.0


#  **Ridge Regression**
Ridge extends LinearRegression by providing L2 regularization on the coefficients when predicting response y with a linear combination of the predictors in X. It can reduce the variance of the predictors, and improves the conditioning of the problem.

The Ridge Regression function implemented in the cuml library allows the user to change the fit_intercept, normalize, solver and alpha parameters. Here is a brief on RAPIDS' Ridge Regression's parameters:
1. alpha:float or double. Regularization strength - must be a positive float. Larger values specify stronger regularization. Array input will be supported later.
1. solver:‘eig’ or ‘svd’ or ‘cd’ (default = ‘eig’). Eig uses a eigendecomposition of the covariance matrix, and is much faster. SVD is slower, but is guaranteed to be stable. CD or Coordinate Descent is very fast and is suitable for large problems.
1. fit_intercept:boolean (default = True). If True, Ridge tries to correct for the global mean of y. If False, the model expects that you have centered the data.
1. normalize:boolean (default = False). If True, the predictors in X will be normalized by dividing by it’s L2 norm. If False, no scaling will be done.

The methods that can be used with the Ridge Regression are:
1. fit: Fit the model with X and y.
1. get_params: Sklearn style return parameter state
1. predict: Predicts the y for X.
1. set_params: Sklearn style set parameter state to dictionary of params.

The model accepts only numpy arrays or cudf dataframes as the input. In order to convert your dataset to cudf format please read the cudf documentation on https://rapidsai.github.io/projects/cudf/en/latest/. It is important to understand that the 'svd' solver will run slower than the 'eig' solver however, the 'svd' solver is more stable and robust. Therefore, we would recomend that you use the 'eig' solver when a slight error is acceptable. For additional information please refer to the documentation on https://rapidsai.github.io/projects/cuml/en/latest/index.html

In [0]:
import numpy as np
import pandas as pd
import cudf
import os

from cuml import Ridge as cuRidge
from sklearn.linear_model import Ridge as skRidge

from sklearn.datasets import make_regression
from sklearn.metrics import mean_squared_error

# Select a particular GPU to run the notebook  
os.environ["CUDA_VISIBLE_DEVICES"]="0"

In [0]:
from google.colab import files
uploaded = files.upload()

Saving mortgage.npy.gz to mortgage.npy.gz


## Loading Data

In [0]:
# check if mortgage dataset is present and then extract the data from it, else just create a random dataset for regression
import gzip
# change the path of the mortgage dataset if you have saved it in a different directory
def load_data(nrows, ncols, cached = 'mortgage.npy.gz'):
    train_rows = int(nrows*0.8)
    if os.path.exists(cached):
        print('use mortgage data')

        with gzip.open(cached) as f:
            X = np.load(f)
        # the 4th column is 'adj_remaining_months_to_maturity'
        # used as the label
        X = X[:,[i for i in range(X.shape[1]) if i!=4]]
        y = X[:,4:5]
        rindices = np.random.randint(0,X.shape[0]-1,nrows)
        X = X[rindices,:ncols]
        y = y[rindices]
        df_y_train = pd.DataFrame({'fea%d'%i:y[0:train_rows,i] for i in range(y.shape[1])})
        df_y_test = pd.DataFrame({'fea%d'%i:y[train_rows:,i] for i in range(y.shape[1])})
    else:
        print('use random data')
        # create a random regression dataset
        X,y = make_regression(n_samples=nrows,n_features=ncols,n_informative=ncols, random_state=0)
        df_y_train = pd.DataFrame({'fea0':y[0:train_rows,]})
        df_y_test = pd.DataFrame({'fea0':y[train_rows:,]})

    df_X_train = pd.DataFrame({'fea%d'%i:X[0:train_rows,i] for i in range(X.shape[1])})
    df_X_test = pd.DataFrame({'fea%d'%i:X[train_rows:,i] for i in range(X.shape[1])})

    return df_X_train, df_X_test, df_y_train, df_y_test

In [0]:
%%time
# nrows = number of samples
# ncols = number of features of each sample 

nrows = 2**20 
ncols = 399

#split the dataset into training and testing sets, in the ratio of 80:20 respectively
X_train, X_test, y_train, y_test = load_data(nrows,ncols)

use mortgage data
CPU times: user 14 s, sys: 4.85 s, total: 18.8 s
Wall time: 18.9 s


In [0]:
print('training data',X_train.shape)
print('training label',y_train.shape)
print('testing data',X_test.shape)
print('testing label',y_test.shape)

training data (838860, 399)
training label (838860, 1)
testing data (209716, 399)
testing label (209716, 1)


In [0]:
X_train

fea0      fea1      fea2      fea3  ...  fea395  fea396  fea397  fea398
0       0.736111  0.144390  0.081081  0.730469  ...     0.0     0.0     0.0     0.0
1       0.694444  0.196900  0.018018  0.757812  ...     0.0     0.0     0.0     0.0
2       0.500000  0.335654  0.112613  0.365234  ...     0.0     0.0     0.0     0.0
3       0.527778  0.212032  0.139640  0.705078  ...     0.0     0.0     0.0     0.0
4       0.527778  0.108215  0.189189  0.449219  ...     0.0     0.0     0.0     0.0
...          ...       ...       ...       ...  ...     ...     ...     ...     ...
838855  0.527778  0.311932  0.040541  0.748047  ...     0.0     0.0     0.0     0.0
838856  0.550000  0.204954  0.040541  0.748047  ...     0.0     0.0     0.0     0.0
838857  0.472222  0.023533  0.765766  0.082031  ...     0.0     0.0     0.0     0.0
838858  0.763889  0.196900  0.031532  0.634766  ...     0.0     0.0     0.0     0.0
838859  0.597222  0.196900  0.004505  0.412109  ...     0.0     0.0     0.0     0.0

[838860 rows x 399 columns]

In [0]:
y_train

fea0
0       0.452754
1       0.620426
2       0.513470
3       0.765179
4       0.770004
...          ...
838855  0.836751
838856  0.406514
838857  0.834339
838858  0.822276
838859  0.000000

[838860 rows x 1 columns]

In [0]:
X_test

fea0      fea1      fea2      fea3  ...  fea395  fea396  fea397  fea398
0       0.569444  0.304975  0.099099  0.722656  ...     0.0     0.0     0.0     0.0
1       0.538889  0.341714  0.360360  0.609375  ...     0.0     0.0     0.0     0.0
2       0.727778  0.196900  0.018018  0.757812  ...     0.0     0.0     0.0     0.0
3       0.416667  0.151791  0.337838  0.267578  ...     0.0     0.0     0.0     0.0
4       0.722222  0.035794  0.274775  0.294922  ...     0.0     0.0     0.0     0.0
...          ...       ...       ...       ...  ...     ...     ...     ...     ...
209711  0.569444  0.458277  0.036036  0.750000  ...     0.0     0.0     0.0     0.0
209712  0.638889  0.196900  0.013514  0.525391  ...     0.0     0.0     0.0     0.0
209713  0.638889  0.144282  0.472973  0.560547  ...     0.0     0.0     0.0     0.0
209714  0.527778  0.063477  0.121622  0.361328  ...     0.0     0.0     0.0     0.0
209715  0.763889  0.237233  0.040541  0.748047  ...     0.0     0.0     0.0     0.0

[209716 rows x 399 columns]

In [0]:
y_test

fea0
0       0.513470
1       0.526337
2       0.927624
3       0.659027
4       0.531162
...          ...
209711  0.659831
209712  0.000000
209713  0.242461
209714  0.341375
209715  0.998392

[209716 rows x 1 columns]

####  Scikit-Learn Implementation

In [0]:
%%time
# use the sklearn ridge regression model to fit the training dataset 
skridge = skRidge(fit_intercept = False,
                  normalize = True, 
                  alpha = 0.1)

skridge.fit(X_train, y_train)

CPU times: user 1min 2s, sys: 3.56 s, total: 1min 6s
Wall time: 36.1 s


In [0]:
%%time
# calculate the mean squared error of the sklearn ridge regression model on the testing dataset
sk_predict = skridge.predict(X_test)
error_sk = mean_squared_error(y_test, sk_predict)

CPU times: user 139 ms, sys: 4.2 ms, total: 143 ms
Wall time: 121 ms


####  **cuML Implementation**

In [0]:
%%time
# convert the pandas dataframe to cudf format
X_cudf = cudf.DataFrame.from_pandas(X_train)
X_cudf_test = cudf.DataFrame.from_pandas(X_test)

y_cudf = y_train.values
y_cudf = y_cudf[:,0]
y_cudf = cudf.Series(y_cudf)

y_cudf_test = y_test.values
y_cudf_test = y_cudf_test[:,0]
y_cudf_test = cudf.Series(y_cudf_test)

CPU times: user 1.21 s, sys: 224 ms, total: 1.43 s
Wall time: 1.47 s


In [0]:
X_cudf.shape

(838860, 399)

In [0]:
X_cudf_test.shape

(209716, 399)

In [0]:
y_cudf_test.shape

(209716,)

CANNOT TRAIN ON 838860 SIZED DATA, RUNTIME CRASHED DUE TO OUT OF MEMORY, USING A SMALL SUBSET TO DEMO

In [0]:
X_cudf_subset = X_cudf[:10000]
y_cudf_subset = y_cudf[:10000]
X_cudf_test_subset = X_cudf_test[:2000]
y_cudf_test_subset = y_cudf_test[:2000]

In [0]:
X_cudf_subset

fea0      fea1      fea2      fea3  ...  fea395  fea396  fea397  fea398
0     0.736111  0.144390  0.081081  0.730469  ...     0.0     0.0     0.0     0.0
1     0.694444  0.196900  0.018018  0.757812  ...     0.0     0.0     0.0     0.0
2     0.500000  0.335654  0.112613  0.365234  ...     0.0     0.0     0.0     0.0
3     0.527778  0.212032  0.139640  0.705078  ...     0.0     0.0     0.0     0.0
4     0.527778  0.108215  0.189189  0.449219  ...     0.0     0.0     0.0     0.0
...        ...       ...       ...       ...  ...     ...     ...     ...     ...
9995  0.555556  0.196900  0.018018  0.757812  ...     0.0     0.0     0.0     0.0
9996  0.555556  0.040701  0.531532  0.183594  ...     0.0     0.0     0.0     0.0
9997  0.597222  0.457806  0.788288  0.423828  ...     0.0     0.0     0.0     0.0
9998  0.666667  0.196900  0.013514  0.759766  ...     0.0     0.0     0.0     0.0
9999  0.541667  0.019646  0.738739  0.093750  ...     0.0     0.0     0.0     0.0

[10000 rows x 399 columns]

In [0]:
%%time
# run the cuml ridge regression model to fit the training dataset.  Eig is the faster algorithm, but svd is more accurate 
curidge = cuRidge(fit_intercept = False,
                  normalize = True,
                  solver = 'svd', 
                  alpha = 0.1)

curidge.fit(X_cudf_subset, y_cudf_subset)

CPU times: user 302 ms, sys: 53.9 ms, total: 356 ms
Wall time: 357 ms


In [0]:
%%time
# calculate the mean squared error of the testing dataset using the cuml ridge regression model
cu_predict = curidge.predict(X_cudf_subset).to_array()
error_cu = mean_squared_error(y_cudf_subset, cu_predict)

CPU times: user 3.49 s, sys: 65.6 ms, total: 3.55 s
Wall time: 3.56 s


In [0]:
# print the mean squared error of the sklearn and cuml model to analyse them
print("SKL MSE(y):", error_sk)
print("CUML MSE(y):", error_cu)

SKL MSE(y): 2.9886542e-13
CUML MSE(y): 1.2581127e-09


Note: Above just try to make the process complete, since used subset of dataset, cannot compare the difference between SK-learn & cuML.